kube-proxy local traffic detector single-vs-dual-stack cleanup

danwinship opened the pull request:
#### What type of PR is this?
/kind cleanup
/sig network

#### What this PR does / why we need it:

    Previously the single-stack and dual-stack modes of kube-proxy had
    completely separate LocalDetector-creation code paths, with different
    error-handling behavior (where the single-stack mode only looked at
    the legacy single-stack config fields and errored out if they had the
    wrong family, while the dual-stack mode looked at the dual-stack
    config fields, allowed them to be in either order, and fell back to
    no-op when no config was available for a family).
    
    Given that the "dual-stack" mode is kube-proxy's normal mode, and
    "single-stack" mode is mostly just a hack to avoid printing error
    messages when IPv6 is disabled in the kernel, it doesn't make sense
    for errors to be handled differently in the two modes. So make the
    single-stack mode work the same way as the dual-stack mode.

from #117436. (Making the dual-stack local detector constructor be just "call the single-stack local detector constructor for IPv4 then call the single-stack local detector constructor for IPv6" is a pre-req for refactoring the single-vs-dual-stack mode split out into shared code.)

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

danwinship committed:
Simplify getDualStackLocalDetectorTuple

Since the single-stack and dual-stack local-detector-getters now have
the same behavior in terms of error-checking and dual-stack config, we
can just replace the contents of getDualStackLocalDetectorTuple() with
a pair of calls to getLocalDetector().

danwinship committed:
Remove unnecessary utiliptables.Interface arg from local detectors

getLocalDetector() used to pass a utiliptables.Interface to
NewDetectLocalByCIDR() so that NewDetectLocalByCIDR() could verify
that the passed-in CIDR was of the same family as the iptables
interface. It would make more sense for getLocalDetector() to verify
this itself and just *not call NewDetectLocalByCIDR* if the families
don't match, and that's what the code does now. So there's no longer
any need to pass the utiliptables.Interface to the local detector.

dims commented:
/assign @thockin @aojea

aojea commented on cmd/kube-proxy/app/server_others_test.go:
```diff
@@ -278,17 +278,17 @@ func Test_getLocalDetector(t *testing.T) {
+			expected:    proxyutiliptables.NewNoOpLocalDetector(),
 			nodeInfo:    makeNodeWithPodCIDRs("2002::1234:abcd:ffff:c0a8:101/96"),
-			errExpected: true,
+			errExpected: false,
```
why should not we fail? this config will not work well

danwinship replied:
To be clear, "this config" is "you have a host where IPv6 has been disabled at the kernel level, and then you try to add that host to a single-stack IPv6 cluster". And yeah, it will not work well. In fact, it will work so badly, I'm not convinced you could actually have gotten far enough to reach this error message...

If we really want to check this sort of stuff, then we should check _all_ of the relevant configuration ahead of time, and then fall back to single-stack mode if _any_ piece of configuration requires it, and error out if any two pieces of configuration are incompatible.

danwinship replied:
(So, just to be clear: 99% of clusters run kube-proxy in dual-stack mode, because it only falls back to single-stack mode if IPv6 is disabled in the kernel. So changing the expected result of this test has no effect on 99% of users.)

aojea replied:
> single-stack mode if IPv6 is disabled in the kernel

I'm not sure the percentage is so great, some environments disable ipv6 for security reasons ... or for reasons

I think we should keep failing, maybe in another place we can validate and not here, but I don't think that being right 99% of times when we can be 100% is a good argument

aojea commented on cmd/kube-proxy/app/server_others.go:
```diff
@@ -393,123 +393,58 @@ func detectNumCPU() int {
 	case proxyconfigapi.LocalModeInterfaceNamePrefix:
 		return proxyutiliptables.NewDetectLocalByInterfaceNamePrefix(config.DetectLocal.InterfaceNamePrefix)
 	}
-	klog.InfoS("Defaulting to no-op detect-local", "detectLocalMode", string(mode))
```
don't you find useful to see why get here? you can get from teh config the value but having it in one line saves time

danwinship replied:
in both cases where you can end up here, we logged the configured mode immediately before:

`klog.InfoS("Detect-local-mode set to ClusterCIDR, but no cluster CIDR for family", "ipFamily", ipFamily)`
`klog.InfoS("Detect-local-mode set to NodeCIDR, but no PodCIDR defined at node for family", "ipFamily", ipFamily)`

aojea replied:
oh, didn't realize

aojea commented:
nice refactor LGTM once the two comments I left are solved, I think we should not loose validation despite how improbable it is, the other comment is a nit about logging, not going to fight that last one 😄

danwinship commented:
/hold
will do the config IP family validation as a separate PR

danwinship commented:
will rebase on#119003 after that merges

thockin commented on cmd/kube-proxy/app/server_others.go:
```diff
@@ -435,15 +430,15 @@ func getLocalDetector(mode proxyconfigapi.LocalMode, config *proxyconfigapi.Kube
 	var err error
 
-	localDetectors[0], err = getLocalDetector(mode, config, ipt[0], nodeInfo)
+	localDetectors[0], err = getLocalDetector(v1.IPv4Protocol, mode, config, nodeInfo)
```
IIUC this makes the ordering of v4,v6 fixed, right?  If so, would it make more sense to return a `map[IPFamily]LocalTrafficDetector` or even a struct with `ipv4` and `ipv6` fields, rather than an obscure array?

danwinship replied:
It was always that way. Yes, it's weird. It will go away in the next PR (because we won't ever keep track of two LocalTrafficDetectors at once any more; we'll create an IPv4 LocalDetector when/if we create an IPv4 proxier, and we'll create an IPv6 LocalDetector when/if we create an IPv6 proxier).

thockin reviewed (commented):
Good cleanup

danwinship committed:
Improve the single-stack LocalDetector behavior

1. When bringing up a single-stack kube-proxy in a dual-stack cluster,
allow using either the primary or secondary IP family.

2. Since the earlier config-checking code will already have bailed out
if the single-stack configuration is unusably broken, we don't need to
do that here. Instead, just return a no-op local detector if there are
no usable CIDRs of the expected IP family.

danwinship committed:
Add names to the getLocalDetector unit tests, use t.Run

danwinship commented:
/hold cancel

danwinship commented:
/test unit

danwinship commented:
/test pull-kubernetes-unit

thockin commented:
Thanks!

/lgtm
/approve
